### ISSUE TITLE
Create Pipelines

### ISSUE THREAD

**Issue by dati91** [szeged/webrender#8]
Define the correct pipeline, vertex, etc. definitions.

They depends on the shaders and there are 3 type of shaders: primitive(simple&transform), cache, chache_clip.

Our current approach is to implement one of them (ps_rectangle), then implement the rest.

**Comment by dati91** [comment]
The current status: https://github.com/dati91/webrender/commits/gfx_device

We will have a minimized `ps_rectangle` shader with the necessary fields.
Then try to provide the correct texture (data16, layer, render tasks, etc.) and instanced (indexes, transform, position) data and draw a simple rectangle with it.

We can get the texture (shader sampler2d) data from the current frame in a draw call. (e.g. `frame.gpu_data16`).
The instanced data is in the `PrimitiveInstance` structure, which we can get from the proper render task.

**Comment by dati91** [comment]
I was able to draw a rectangle with the gfx backend: http://imgur.com/a/V83GJ (without the red lines!)
It is a very early stage and works only with [this example](https://github.com/dati91/webrender/blob/gfx_device/rect_sample/src/main.rs).
I had some problem with the different coordinate system, the correct image should look like [this](http://imgur.com/a/jKXKX), so I will have to do some translation later (and the other backends will differ as well).

**Comment by dati91** [comment]
I managed to draw a rectangle correctly: [rect_sample_gfx](http://imgur.com/a/GVVQ2) vs [rect_sample_orig](http://imgur.com/a/6DOAg)
Only the rectangle should be visible so that works correctly.
The background color (which is a `clear_color`) still differ... it seems to me that the problem is with the gfx (or i am doing something terribly wrong, which is a real possibility). The color should be `[0.3, 0.0, 0.0] RGB(76,0,0)` but the result color (according to gimp) is `[0.58, 0.0, 0.0] RGB(149,0,0)`.

The next step will be to use the whole shader, not just a minimized one, then create a `Program/Shader` structure that bundle everything together. (pso, slice, data, textures/samplers, etc)
Finally extend it with other (primitive) shaders.

**Comment by kvark** [comment]
> The background color (which is a clear_color) still differ... it seems to me that the problem is with the gfx

That is actually correct behavior, given your framebuffer is sRGB. The shader outputs `0.3`, which gets automatically converted into sRGB space by using an inverse of `2.2` power, thus producing `0.3 ^ (1/2.2) ~= 0.5785326`, which is what gets written.

TL;DR: when using gamma-correct rendering, you need to make sure all the colors come in as linear, including the ones you provide in constants or for color clears. In your case, `0.3` should have been powered by a factor of `2.2` before passing to `clear_color`.

**Comment by dati91** [comment]
Ah great, thanks!

**Comment by dati91** [comment]
Servo with the gfx-ps_rectange code: http://imgur.com/a/E820D (It scroll-able as well)
